rhealpixdggs/rhp_wrappers.py

```diff
@@ -0,0 +1,538 @@
+
+        # We're done if k_eff takes us all the way to the opposite cell (shouldn't happen at this point...)
+        if k_eff < 1:
+            ring.append(_suid_to_str(cell.suid))
+
+        # We have to do the full walk around the ring
+        else:
+            # Set starting point
+            cell, direction, n_steps = _find_cell_ring_start(
+                cell, k_eff, max_steps, directions, direction_inverse
+            )
+
+            # Walk around the ring one side at a time and collect cell addresses
+            for _ in range(0, len(directions)):
+                step = 0
+                while step < n_steps:
+                    # Add index to ring, take a step
+                    ring.append(_suid_to_str(cell.suid))
+                    next = cell.neighbor(direction)
+
+                    # Looking back not being the same as looking ahead means we need to realign
+                    if next.neighbor(direction_inverse[direction]) != cell:
+                        direction = direction_inverse[_neighbor_direction(next, cell)]
+
+                    # Take the step
+                    cell = next
+                    step = step + 1
+
+                # Prepare walking direction for next ring side
+                if n_steps == 2 * k_eff:
+                    direction = directions[
+                        (directions.index(direction) + 1) % len(directions)
+                    ]
+
+                # Reset number of steps along a side
+                n_steps = max_steps
+
+    return ring
+
+
+# TODO: placeholder, find out if rhp needs that function
+# def weighted_cell_ring(rphindex: str, k: int = 1) -> list[str]:
+#    pass
+
+
+def k_ring(rhpindex: str, k: int = 1, verbose: bool = True) -> list[str]:
+    """
+    Returns the k-ring of cell indices around rhpindex at distance k (or None if rhpindex is invalid).
+
+    Also returns None if k < 0.
+
+    TODO: give the option to select another predefined DGGS, or pass in a custom one
+    """
+    if verbose:
+        warn(str.format(CELL_RING_WARNING, "k"))
+
+    if not rhp_is_valid(rhpindex) or (k < 0):
+        return None
+
+    # A k-ring at distance 0 just consists of the centre cell itself
+    if k == 0:
+        return [rhpindex]
+
+    distance = min(2 * WGS84_003.N_side ** rhp_get_resolution(rhpindex), k)
+    kring = [rhpindex]
+
+    for d in range(1, distance + 1):
+        kring = kring + cell_ring(rhpindex, d, verbose=False)
+
+    return kring
+
+
+# TODO: placeholder, find out if rhp needs that function
+# def k_ring_smoothing(rhpindex: str, k: int = 1) -> list[str]:
+#    pass
+
+
+def polyfill(geometry, resolution: int) -> list[str]:
+    raise NotImplementedError()
+
+
+def linetrace(geometry, resolution: int) -> list[str]:
+    raise NotImplementedError()
+
+
+# ======== Helper functions ======== #
+
+
+def _suid_to_str(suid: tuple) -> str:
+    return "".join([str(d) for d in suid])
```

Comment:
The implementation differs from the `Cell.__str__` method, is that expected? It also seems to be used as `_suid_to_str(cell)`; would rather `str(cell)` be better?

Reply:
It's intentional that they're different. Cell.__str__ returns results that are not consistent across resolutions, which adds unnecessary complexity down the line. I feel that it's best to keep Cell.__str__ for display/print purposes only.

Passing the entire Cell object to _suid_to_str is worse from an encapsulation perspective: we'd be passing in a heavy object when we only need one tuple from its properties. The best practice that I learned is to separate out the tuple so the function doesn't need to know about the Cell class as a whole. (That also means that e.g. the dggs object could pass in an appropriate tuple id and have it converted to string - if that's something we need in the future.)

Reply:
> Passing the entire Cell object to _suid_to_str

I'm not recommending this; but it does seem odd looking in that there's a cell `__str__` method which does almost the same thing, but we don't use it in favour of this private function. The difference is only apparent when the condition `(self.rdggs.N_side) ** 2 < 10` is met, and given we're only using `N_side = 3`, this condition won't be met. Are we therefore better off using `Cell.__str__` for the sake of consistency?

Reply:
We can use that if you're happy to make the changes to the wrapper. I'm a bit out of hours to dive into it atm.

Reply:
I've removed the function in favour of using Cell.__str__; no other changes needed and all tests pass.